collectors/htcondor/collector.py

```diff
@@ -0,0 +1,242 @@
+
+import logging
+
+from subprocess import Popen, PIPE
+from datetime import datetime as dt
+from typing import Union, Tuple, List
+from urllib.parse import quote
+
+from pyauditor import (
+    AuditorClientBuilder,
+    AuditorClient,
+    Record,
+    Component,
+    Score,
+    Meta,
+)
+
+
+from utils import maybe_convert, get_value
+from config import Config
+from state_db import StateDB
+
+
+class CondorHistoryCollector(object):
+    def __init__(self, config: Config):
+        self.config = config
+        self.logger = self.setup_logger()
+        self.client = self.setup_auditor_client()
+        self.state_db = StateDB(config.state_db)
+
+    def setup_auditor_client(self) -> AuditorClient:
+        """Sets up the auditor client."""
+        builder = AuditorClientBuilder().address(self.config.addr, self.config.port)
+        return builder.build()
+
+    def setup_logger(self) -> logging.Logger:
+        """Sets up the logger for the collector."""
+        logger = logging.getLogger("auditor.collectors.htcondor")
+        logger.setLevel(self.config.log_level)
+        if self.config.log_file:
+            from logging.handlers import RotatingFileHandler
+
+            handler = RotatingFileHandler(
+                self.config.log_file,
+                maxBytes=10 * 1024 * 1024,
+                backupCount=5,
+            )
+        else:
+            handler = logging.StreamHandler()
+        handler.setFormatter(
+            logging.Formatter(
+                "{asctime} - {name} - {levelname: <8} - {message}", style="{"
+            )
+        )
+        logger.addHandler(handler)
+        return logger
+
+    async def run(self):
+        self.logger.info("Starting collector run.")
+        with self.state_db.connection():
+            for schedd_name in self.config.schedd_names:
+                id = self.config.get("job_id") or ""
+                await self._collect(schedd_name, job_id=id)
+        self.logger.info("Collector run finished.")
+
+    async def _collect(self, schedd_name: str, job_id: str = ""):
+        """Collects jobs from `condor_history` for a given schedd and adds them to the auditor client.
+
+        Jobs are iterated over in reverse order, so that the oldest job is processed first.
+        """
+        self.logger.info(f"Collecting jobs for schedd {schedd_name}.")
+        if job_id:
+            try:
+                if "." in job_id:
+                    cluster, proc = map(int, job_id.split("."))
+                else:
+                    cluster, proc = int(job_id), 0
+            except ValueError:
+                raise ValueError("Invalid job id.")
+        else:
+            cluster, proc = self.get_last_job(schedd_name)
+        self.logger.debug(f"Using job id {cluster}.{proc}.")
+
+        jobs = self.query_htcondor_history(schedd_name, cluster, proc)
+
+        added, failed = 0, 0
+        for job in jobs[::-1]:
+            if record := self._generate_record(job):
```

Comment:
Support for Python 3.6 (!) would be great, since that is still the default on RHEL7 run by many grid sites.

This would be easier if `_generate_record` would raise an exception if no record can be generated.

Reply:
Scratch 3.6. I only checked the `pyproject.toml` but the Maturin build backend already dropped support for 3.6. Targeting 3.8 should be fine, then. The distributions I'm aware of directly went from 3.6 to 3.8, plus 3.7 will be EOL soon as pointed out already.

I'd still recommend handling this via exceptions instead of `None` (at the very least, check for `is None`) but in principle both is fine.

Reply:
Thanks @dirksammel and @maxfischer2781 for your comments. I've removed the use of the walrus operator, even though `pyauditor` does not work with Python 3.6 as @maxfischer2781 pointed out (I've created a ticket for that: #267).
I've also introduced a custom exception which is thrown if no record can be generated.

Reply:
> Scratch 3.6. I only checked the pyproject.toml but the Maturin build backend already dropped support for 3.6. 

We're using an older version of PyO3 in order to be able to build for Python 3.6 (since TARDIS requires this too). Therefore I believe it should be possible to support 3.6, but there is still the question whether it is worth the effort for this collector.